### DISCUSSION TITLE
Cross OS fish shell function for package management

### DISCUSSION THREAD

**Discussion by MANICX100** [fish-shell/fish-shell#9163] [Q&A] [answered]
I am trying to have functions that work across distributions to install, remove or cleanup the chosen package managers (AUR in case of arch) namely dnf fedora, nala debian and paru Arch
below i am getting the id of the distribution
   
   alias getos="grep -ioP '^ID=\K.+' /etc/os-release"

and want to do something like

    alias getos="grep -ioP '^ID=\K.+' /etc/os-release"
    set osinfo getos

    function u
	    switch $osinfo
	        case fedora
		    sudo dnf install
	        case arch
		    paru -S
	        case debian
		    sudo nala install
	        case '*'
		    echo 'fail'
	    end
    end
    
    
    function r
	    switch $osinfo
	        case fedora
		    sudo dnf remove
	        case arch
		    paru -Rns
	        case debian
		    sudo nala autoremove
	        case '*'
		    echo 'fail'
	    end
    end
    
    
    function c
	    switch $osinfo
	        case fedora
		    sudo dnf clean all
		    flatpak uninstall --unused
	        case arch
		    paru -Yc
		    flatpak uninstall --unused
	        case debian
		    sudo nala autoremove -o APT::Autoremove::RecommendsImportant=0 -o APT::Autoremove::SuggestsImportant=0
		    flatpak uninstall --unused
	        case '*'
		    echo 'fail'
	    end
    end

**Comment by ajgringo619** [comment]
I would change (2) things:

1. Use longer function names (not a fan of one-character aliases/functions)
2. Separate the long multiple function lines. For example, in 'function c':
```
case debian
    sudo nala autoremove -o APT::Autoremove::RecommendsImportant=0 -o APT::Autoremove::SuggestsImportant=0
    flatpak uninstall --unused
```
Just makes it easier to read.

**Reply by MANICX100** [reply]
Good idea thanks.

Any idea why my switch statements always evaluate to fail
?

I have updated the OP

**Reply by faho** [reply]
>Any idea why my switch statements always evaluate to fail

```fish
alias getos="grep -ioP '^ID=\K.+' /etc/os-release"
set osinfo getos
```

This... creates a variable "osinfo" and sets it to the string "getos". It doesn't *run* the function.

You want:

```fish
alias getos="grep -ioP '^ID=\K.+' /etc/os-release"
set -g osinfo (getos)
```

or, tbh

```fish
set -g osinfo (grep -ioP '^ID=\K.+' /etc/os-release)
```

The `-g` makes it specifically create a *global* variable. If you didn't pass that, it could use a pre-existing variable instead.

**Reply by MANICX100** [reply]
thank you

**Reply by MANICX100** [reply]
Hmmmm now it just executes without the arguments provided,

I have tried both

    sudo dnf install $1

and preferably

    sudo dnf install $argv

all evaluate as just sudo dnf install

**Reply by faho** [reply]
`$1` isn't a thing in fish.

You want `$argv[1]`, or indeed `$argv`.

And then, well, you'll actually have to run `u foo`. If you run it elsewhere, you'll have to pass on the arguments there, like running `u $argv`.

Most likely you've made a mistake in the *setup* - either you're executing the function from elsewhere, or it's already loaded and your shell still has the old definition without $argv. You haven't provided that information.

**Reply by MANICX100** [reply]
Thank you $argv did indeed work. I was running a script to automatically grab changes to my config from github but it had a 300 second cache.

**Comment by MANICX100** [comment, marked as answer]
@faho - thank you for the answer and further assistance.

> > Any idea why my switch statements always evaluate to fail
> 
> ```fish
> alias getos="grep -ioP '^ID=\K.+' /etc/os-release"
> set osinfo getos
> ```
> 
> This... creates a variable "osinfo" and sets it to the string "getos". It doesn't _run_ the function.
> 
> You want:
> 
> ```fish
> alias getos="grep -ioP '^ID=\K.+' /etc/os-release"
> set -g osinfo (getos)
> ```
> 
> or, tbh
> 
> ```fish
> set -g osinfo (grep -ioP '^ID=\K.+' /etc/os-release)
> ```
> 
> The `-g` makes it specifically create a _global_ variable. If you didn't pass that, it could use a pre-existing variable instead.
